rel: Release v2.0.0 (#743)
honeycombio/refinery: RELEASE_NOTES.md

While [CHANGELOG.md](./CHANGELOG.md) contains detailed documentation and links to all of the source code changes in a given release, this document is intended to be aimed at a more comprehensible version of the contents of the release from the point of view of users of Refinery.

## Version 2.0.0

> This is a major release of Refinery with breaking changes to configuration files, sampling rules, and operations.

Reviewer: ```suggestion
This is a major release of Refinery with a new configuration file format, new samplers, and updated runtime behavior. It has several breaking changes, and so it also comes with a conversion tool.
```

### Configuration File Changes

The configuration and sampler file formats hav been completely redesigned. Key changes:

- The preferred file format is now YAML instead of TOML. All examples and documentation now use YAML format (the old *_complete.toml files now have 1.x in their names).
- The config file is now organized into sections for clarity.

> - Sampler rules now require default target named `__default__` to specify the default behavior of sampling.

Reviewer: ```suggestion
- Sampler rules now require default target named `__default__` to specify the default sampling behavior.
```

- Many default values have changed to be more useful.

> - Configurations are now fully validated - misspellings, type errors, faulty indentions, and extra values are now detected.

Reviewer: ```suggestion
- Configurations are now fully validated - misspellings, type errors, faulty indentions, and extra values are now detected. Refinery will no longer run if your configurations are invalid.
```

- Documentation for configuration is now automatically generated so that it will stay in sync with the source.

> - A conversion tool has been provided to convert a v1 file to the new format.

Reviewer: ```suggestion
- A conversion tool has been provided to convert a v1.x configuration file to the new format.
```

Specific configuration changes worth noting:

> - The configuration version is a required field; this is to permit future configuration format changes without breaking existing configurations again.

Reviewer: ```suggestion
- There is now a required configuration version field; this is to permit future configuration format changes without breaking existing configurations again.
```

- All duration values like timeouts, tickers, and delays are now specified as durations like `5s`, `1m30s`, or `100ms`.
- Memory sizes can now have a standard suffix like `MiB` or `GB`.
- Instead of calculating a maximum memory usage value, it is now possible to specify `AvailableMemory` as the total memory available, plus `MaxMemoryPercentage`.
- Some legacy operational controls relating to caching and memory management have been removed.
- Refinery can be run with a validation command-line switch (`-V`), which can be used to validate configuration (typically used for CI/CD).
- StressRelief activation level defaults have been changed to higher values.
- Log levels have been cleaned up so that they're more useful. The `info` level is less noisy and `warn` is now a valid option.
- Config file location can also be a URL instead of a file path. The contents are retrieved with a simple GET request.
- Config file contents are periodically refetched at the rate specified by `ConfigReloadInterval`. An immediate reload can be forced by sending the `SIGUSR1` signal.

### Sampler Changes

> - All dynamic samplers now correctly count spans, not traces. Although they were documented as counting the number of spans, in fact they were only counting traces, which often made it difficult to achieve appropriate target rates. *** After running the conversion tool, existing configurations should be adjusted! ***

Reviewer: ```suggestion
- All dynamic samplers now correctly count spans, not traces. Although they were documented as counting the number of spans in previous versions, they were in fact only counting traces, which often made it difficult to achieve appropriate target rates.*** After running the conversion tool, existing configurations should be adjusted! ***
```

> - New Samplers: The WindowedThroughputSampler and EMAThroughputSampler both use dynamic sampling techniques to adjust sample rates to achieve a desired throughput. The WindowedThroughputSampler does so with a moving window of samples, while the EMAThroughputSampler maintains a moving average.

Reviewer: ```suggestion
- New Samplers: The `WindowedThroughputSampler` and `EMAThroughputSampler`  are two new samplers. We highly recommend replacing any use of TotalThroughputSampler in favor of one of these. Both use dynamic sampling techniques to adjust sample rates to achieve a desired throughput. The `WindowedThroughputSampler` does so with a moving window of samples, while the `EMAThroughputSampler` maintains a moving average.
```

Should we also recommend people move off of EMADynamicSampler?

- Individual samplers now report metrics relating to key size and the number of spans and traces processed.

> - Samplers now always have a bounded MaxKeys value, which defaults to 500. Systems relying on a larger keyspace should set this value explicitly for a sampler.

Reviewer: ```suggestion
- Samplers now always have a bounded `MaxKeys` value, which defaults to `500`. Systems relying on a larger keyspace should set this value explicitly for a sampler.
```

### Refinery Metrics Updates

> - Sending metrics with Open Telemetry is now supported, and preferred over Refinery's legacy metrics.

Reviewer: ```suggestion
- Sending metrics with OpenTelemetry is now supported, and preferred over now legacy Honeycomb metrics.
```

> - Refinery's metrics can now be sent to more than one destination (for example, both Prometheus and Open Telemetry).

Reviewer: ```suggestion
- Refinery's metrics can now be sent to more than one destination (for example, both Prometheus and OpenTelemetry).
```

While [CHANGELOG.md](./CHANGELOG.md) contains detailed documentation and links to all of the source code changes in a given release, this document is intended to be aimed at a more comprehensible version of the contents of the release from the point of view of users of Refinery.

## Version 2.0.0

> This is a major release of Refinery with a new configuration file format, new samplers, and updated runtime behavior. It has several breaking changes, and so it also comes with a conversion tool.

Reviewer 2: @TylerHelmuth Would you want to put an internal link to the Conversion Tool section here?

Reviewer 3: Good idea

Reviewer 3: Added

> This is a major release of Refinery with a new configuration file format, new samplers, and updated runtime behavior. It has several breaking changes, and so it also comes with a conversion tool.

Reviewer 2: ```suggestion
This is a major release of Refinery with a new configuration file format, new samplers, and updated runtime behavior. It has several breaking changes for configuration and sampling, and so it also comes with a conversion tool.
```

Reviewer 3: Added the word `especially`, but I think I captured the same sentiment as the suggestion.  I didn't want to make it read like the only breaking changes were configuration and sampling.

- All duration values like timeouts, tickers, and delays are now specified as durations like `5s`, `1m30s`, or `100ms`.
- Memory sizes can now have a standard suffix like `MiB` or `GB`.
- Instead of calculating a maximum memory usage value, it is now possible to specify `AvailableMemory` as the total memory available, plus `MaxMemoryPercentage`.
- Some legacy operational controls relating to caching and memory management have been removed.
- Refinery can be run with a validation command-line switch (`-V`), which can be used to validate configuration (typically used for CI/CD).
- StressRelief activation level defaults have been changed to higher values.
- Log levels have been cleaned up so that they're more useful. The `info` level is less noisy and `warn` is now a valid option.
- Config file location can also be a URL instead of a file path. The contents are retrieved with a simple GET request.
- Config file contents are periodically refetched at the rate specified by `ConfigReloadInterval`. An immediate reload can be forced by sending the `SIGUSR1` signal.

> ### Sampler Changes

Reviewer 2: ```suggestion
### Sampler Changes

```

> - All dynamic samplers now correctly count spans, not traces. Although they were documented as counting the number of spans in previous versions, they were in fact only counting traces, which often made it difficult to achieve appropriate target rates.*** After running the conversion tool, existing configurations should be adjusted! ***

Reviewer 2: ```suggestion
- All dynamic samplers now correctly count spans, not traces. Although they were documented as counting the number of spans in previous versions, they were in fact only counting traces, which often made it difficult to achieve appropriate target rates. **After running the conversion tool, existing configurations should be adjusted!**
```

> - New Samplers: The `WindowedThroughputSampler` and `EMAThroughputSampler`  are two new samplers. We highly recommend replacing any use of TotalThroughputSampler in favor of one of these. Both use dynamic sampling techniques to adjust sample rates to achieve a desired throughput. The `WindowedThroughputSampler` does so with a moving window of samples, while the `EMAThroughputSampler` maintains a moving average.

Reviewer 2: ```suggestion
- New Samplers: The `WindowedThroughputSampler` and `EMAThroughputSampler`  are two new samplers. We highly recommend replacing any use of `TotalThroughputSampler` in favor of one of these. Both use dynamic sampling techniques to adjust sample rates to achieve a desired throughput. The `WindowedThroughputSampler` does so with a moving window of samples, while the `EMAThroughputSampler` maintains a moving average.
```

- Individual samplers now report metrics relating to key size and the number of spans and traces processed.
- Samplers now always have a bounded `MaxKeys` value, which defaults to `500`. Systems relying on a larger keyspace should set this value explicitly for a sampler.

> ### Refinery Metrics Updates

Reviewer 2: ```suggestion
### Refinery Metrics Updates

```

- Sending metrics with OpenTelemetry is now supported, and preferred over now legacy Honeycomb metrics.
- Refinery's metrics can now be sent to more than one destination (for example, both Prometheus and OpenTelemetry).

> ### Notable Bug Fixes

Reviewer 2: ```suggestion
### Notable Bug Fixes

```

While [CHANGELOG.md](./CHANGELOG.md) contains detailed documentation and links to all of the source code changes in a given release, this document is intended to be aimed at a more comprehensible version of the contents of the release from the point of view of users of Refinery.

## Version 2.0.0

This is a major release of Refinery with a new configuration file format, new samplers, and updated runtime behavior. It has several breaking changes, and so it also comes with a conversion tool.

### Configuration File Changes

> The configuration and sampler file formats hav been completely redesigned. Key changes:

Reviewer 2: ```suggestion
The configuration and sampler file formats have been completely redesigned. Key changes:
```

- The preferred file format is now YAML instead of TOML. All examples and documentation now use YAML format (the old *_complete.toml files now have 1.x in their names).
- The config file is now organized into sections for clarity.
- Sampler rules now require default target named `__default__` to specify the default sampling behavior.
- Many default values have changed to be more useful.
- Configurations are now fully validated - misspellings, type errors, faulty indentions, and extra values are now detected. Refinery will no longer run if your configurations are invalid.

> - Documentation for configuration is now automatically generated so that it will stay in sync with the source.

Reviewer 2: @TylerHelmuth Do you want to mention that there's documentation generation for both local and docs site locations?

Reviewer 3: Updated

Reviewer 2: ```suggestion
- Documentation for configuration is now automatically generated so that it will stay in sync with the source. This means that the [Honeycomb docs](https://docs.honeycomb.io/manage-data-volume/refinery/) stay in sync and you can now generate local copies using the [conversion tool](https://github.com/honeycombio/refinery/tree/main/tools/convert).
```

Reviewer 3: Applied.

> - A conversion tool has been provided to convert a v1.x configuration file to the new format.

Reviewer 2: @TylerHelmuth I would actually move the configuration conversion tool bullet point here to being the 2nd bullet point in the list. As a user, reading "You have to use YAML now instead of TOML" would discourage me until I realize there's a conversion tool available.

Reviewer 3: Great idea.

Reviewer 3: Done.

Specific configuration changes worth noting:

- There is now a required configuration version field; this is to permit future configuration format changes without breaking existing configurations again.
- All duration values like timeouts, tickers, and delays are now specified as durations like `5s`, `1m30s`, or `100ms`.
- Memory sizes can now have a standard suffix like `MiB` or `GB`.

> - Instead of calculating a maximum memory usage value, it is now possible to specify `AvailableMemory` as the total memory available, plus `MaxMemoryPercentage`.

Reviewer 2: @TylerHelmuth Not sure what "plus `MaxMemoryPercentage`" means here since this says a max memory usage value is no longer calculated. Are you able to use `MaxMemoryPercentage` as a buffer amount or is the meaning "in addition to specifying"?

Reviewer 3: I'll touch up that sentence to be clearer

Reviewer 3: Updated

Reviewer 2: Looks good!

- Some legacy operational controls relating to caching and memory management have been removed.

> - Refinery can be run with a validation command-line switch (`-V`), which can be used to validate configuration (typically used for CI/CD).

Reviewer 2: @TylerHelmuth Can `--validate` also be used here? (Asking because I only documented the `--validate` option for the Refinery on the docs site.)

Reviewer 3: @mjingle I'm not sure what you're asking

Reviewer 2: Oh! This bullet says that you can use `-V` to validate config. In the Honeycomb docs ([preview specific section here](https://pr1798.prdocs.hny.wtf/manage-data-volume/refinery/migration/#validate)), that switch is described only as `convert validate`. I didn't know if we wanted to reference both options here or not.

Reviewer 3: I think the reference to `convert validate` later in the notes is acceptable.

Reviewer 2: Roger.

### Refinery Metrics Updates

- Sending metrics with OpenTelemetry is now supported, and preferred over now legacy Honeycomb metrics.
- Refinery's metrics can now be sent to more than one destination (for example, both Prometheus and OpenTelemetry).

### Notable Bug Fixes

- Dynamic samplers now count spans, not traces (see above).
- The APIKeys list now applies to OTLP traffic as well as Honeycomb events.
- StressRelief is now more stable and effective.
- Cache overruns should occur much less often and are now a reliable indication that the cache is undersized.

### Conversion Tool

- There is now a `convert` tool included with Refinery -- binary builds are available as part of the release.
- `convert config` can read v1 config files and convert them to v2 YAML files, adding detailed comments.
- `convert rules` can read v1 rules files and convert them to v2 YAML files.
- `convert helm` can read a helm chart with `config` and `rules` sections and write out a revised, updated helm chart.
- `convert validate config` can validate a v2 config file.

> - `convert validate rules` can validate a v2 rules file.

Reviewer 2: @TylerHelmuth Do you want to add the `convert docs` commands for config and rules here? (It's like the only thing that's not listed)

Reviewer 3: Since these are customer-focuses release notes and that command is intended for us I think it is ok to not include it.  That keeps this doc focused on only things the user cares about.  I think calling out that the docs site will be automatically kept up-to-date is the more important fact for the users.

Reviewer 2: Ah, okay. No worries. (We still document this command in both places - both repo and docs site, but I get not needing to highlight it.)
